### PR TITLE
Allow pip to use the local wheel cache while installing dependencies

### PR DESCRIPTION
**Pull request by orf**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

While debugging https://github.com/Homebrew/homebrew-core/pull/117779 locally, I noticed that a significant portion of the time was spent re-installing `setuptools` and other `install dependencies`.

When we run `pip install /path/to/dep`, it will basically always require `setuptools` and `wheel` as a build dependency. Some packages require more, such as `zipp` that needs `packaging-22.0 setuptools-65.6.3 setuptools_scm-7.0.5 tomli-2.0.1 typing-extensions-4.4.0`.

The current code is reinstalling these dependencies over and over again for every single dependency due to the `--no-binary` flag, which is really wasteful.

By passing `--use-feature=no-binary-enable-wheel-cache` it will not download any binaries from pypi but instead it will use the local cached setuptools and wheel dependencies from previous installs:

```
Processing /private/tmp/poetry--requests-20221210-38290-1j8k6b0/requests-2.28.1
  Installing build dependencies: started
  Running command pip subprocess to install build dependencies
  Collecting setuptools>=40.8.0
    Using cached setuptools-65.6.3-py3-none-any.whl (1.2 MB)
  Collecting wheel
    Using cached wheel-0.38.4-py3-none-any.whl (36 kB)
  Installing collected packages: wheel, setuptools
  Successfully installed setuptools-65.6.3 wheel-0.38.4
```

You can see this repeated installation in the [CI output here](https://github.com/Homebrew/homebrew-core/actions/runs/3663040771/jobs/6192604346#step:6:5523).

On my local machine this brings `brew install --verbose --build-bottle poetry --debug` down from 25+ minutes to ~5.

### CODE REVIEW

File: Library/Homebrew/language/python.rb [hunk 1]
```diff
@@ -304,7 +304,7 @@ def pip_install_and_link(targets, link_manpages: false)
         def do_install(targets)
           targets = Array(targets)
           @formula.system @venv_root/"bin/pip", "install",
-                          "-v", "--no-deps", "--no-binary", ":all:",
+                          "-v", "--no-deps", "--use-feature=no-binary-enable-wheel-cache",
```
Comment (carlocab) [inline Library/Homebrew/language/python.rb:307]: Nice. What versions are supported by the `no-binary-enable-wheel-cache` feature?

Comment (orf) [reply Library/Homebrew/language/python.rb:307]: According to this (https://github.com/pypa/pip/pull/11454), 22.3.

This will become the default behaviour in the future, but we can opt into it now.

Comment (carlocab) [reply Library/Homebrew/language/python.rb:307]: Sounds good. We'll need to wait until `python@3.10` is on 22.3+, though. That's in Homebrew/homebrew-core#117507.

**Comment by carlocab** [comment]
Marking as `do not merge` until we have https://github.com/Homebrew/homebrew-core/pull/117507.

File: Library/Homebrew/language/python.rb (hunk 1 above)
Comment (orf) [reply Library/Homebrew/language/python.rb:307]: I wonder - is it possible to bump just `pip`? That should be fairly quick and painless compared to the larger change in https://github.com/Homebrew/homebrew-core/pull/117507.

If this change works as expected, it might also make some of the dependent builds faster I think?

I can make a MR for that if you'd like

Comment (carlocab) [reply Library/Homebrew/language/python.rb:307]: I doubt it'll go much faster that way, TBH. It will need to wait for a long build slot anyway.

**Comment by orf** [comment]
https://github.com/Homebrew/homebrew-core/pull/117507 seems to have been merged, is it possible to give this a try now?

**Comment by branchvincent** [comment]
IIUC, I think we wanted to keep `--no-binary :all:` and just opt-in to the `--use-feature`:

```console
$ pip install --use-feature=no-binary-enable-wheel-cache click
Collecting click
  Downloading click-8.1.3-py3-none-any.whl (96 kB) # <-- binary
...
$ pip install --no-binary :all: --use-feature=no-binary-enable-wheel-cache click
Collecting click
  Downloading click-8.1.3.tar.gz (331 kB)
...
```

**Comment by timle2** [comment]
> This change has caused formulae with pip pinned to a version <22.3 to fail because the `no-binary-enable-wheel-cache` feature does not exist yet.
> 
> ```
> option --use-feature: invalid choice: 'no-binary-enable-wheel-cache' (choose from '2020-resolver', 'fast-deps', 'in-tree-build')
> ```

+1 seeing exactly this after updating brew today

**Comment by orf** [comment]
Damn, is there any reason they where pinned to <22.3?

**Comment by timle2** [comment]
I traced this around for quite awhile trying to figure out why pip 22.0.4 kept showing up in the brew created venvs and subsequently causing formula installs to be blocked. 

Pip was not properly installed in my site packages. That meant venv ran 'ensure-pip' to install pip. The bundled version of pip, bundled with ensurepip is 22.0.4 (I checked https://www.python.org/ftp/python/3.8.16/Python-3.8.16.tar.xz manually for that).

Can't do `pip` as a resource with virtualenv_install_with_resources in the formula to get it updated either, because of the error pip<23 throws (can't pip install pip if pip isn't working).

To resolve, I had to manually install pip 23 into my site-packages, and make sure system_site_packages was enabled for virtualenv_install_with_resources. 

So basically, any blank machines (or machines without site packages setup) will get pip 22.0.4 by default and a broken brew for python formula due to this hardwired `--use-feature=no-binary-enable-wheel-cache` option. 

Also this breaks anyone who is using this setting, since they are 100% getting pip `22.0.4` in their env. 
`virtualenv_install_with_resources(:using => "python@3.8", :system_site_packages => FALSE)`

**Comment by Bo98** [comment]
Since Python 3.9, we modify ensurepip to use a newer pip so the problem doesn't exist on newer Pythons.

This, among other things, haven't been backported to our earlier Pythons. I could do a big backport of changes to 3.8 & 3.7 if needed. I'll need to be careful, but since it's breaking `virtualenv_install_with_resources` it probably makes sense to do.

**Comment by emmyoop** [comment]
@orf We pinned pip because we have been bitten by pip upgrading underneath us in the past and breaking us more than once.

This is what broke: [install definition](https://github.com/dbt-labs/homebrew-dbt/blob/0480fdd42767dfd6c13ec181a4afa33a13a40ef0/.github/formula-template.j2#L21-L25)

And this fixed us: [install definition](https://github.com/dbt-labs/homebrew-dbt/blob/8dc7283699996fbeeb6de358a0820d313efe9275/.github/formula-template.j2#L21-L25)

**Comment by orf** [comment]
Ok, so we have three paths:
1. Revert this 
2. Don't revert this, and set a minimum required pip version for home brew
3. Partially revert this: only enable it for homebrew/homebrew-core CI runs

This is primarily useful for CI build time anyway, so I'd vote for 3?

**Comment by carlocab** [comment]
I'd rather not introduce code paths that never run in CI -- those become too brittle. We're also a rolling release package manager, so requiring a new version of `pip` makes sense to me.

**Comment by MikeMcQuaid** [comment]
> We're also a rolling release package manager, so requiring a new version of `pip` makes sense to me.

Agreed 👍🏻
